Add chapter 7 to document

chiacyu opened the pull request.

chiacyu committed:
Add memory performance tool page

chiacyu committed:
Add memory operation profiling page

chiacyu committed:
Add simulate cpu cache page

chiacyu committed:
Add measure memory usage page

chiacyu committed:
Add improve branch prediction page

chiacyu committed:
Add page fault optimization page

chiacyu committed:
Remove duplicate paragraph

jserv commented on memory-performance-tools/improving-branch-prediction.md:
```diff
@@ -0,0 +1,19 @@
+# 7.4 改善分支預測
+
+6.2.2節中，我們提到了兩個改善 L1i 使用的方法，分別是通過分支預測和重新排序區塊：通過 `__builtin_expect` 進行靜態預測和狀態偵測的最佳化（PGO）。正確的分支預測對軟體的性能有巨大的影響，但在這裡我們會著重在記憶體使用的改善狀況。
```
`6.2.2` 和後方的「節」之間，用一個半形空白區隔

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+了解程式使用了多少記憶體並更進一步知道實際發生的位置，是增進記憶體使用率的第一步。最棒的是目前有一些非常好用的工具可以參考，而且還不需要重新編譯或特別修改程式。
+
+第一個要介紹的工具叫 `massif` ，提供了以以時間為單位計算記憶體使用情況的概觀而且還可以保留編譯器自動產生的 debug 資訊。圖 7.7 為範例。`massif` 與 `cachegrind`（7.2）相同都是使用 `valgrind` 為基礎架構的工具。使用方法如下：
```
提及 massif 時，標註其超連結: https://valgrind.org/docs/manual/ms-manual.html

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+  <figcaption>圖 7.7：Massif統計結果</figcaption>
+</figure>
+
+在進程終止之前，massif 會新增兩個檔案分別為：`massif.XXXXX.txt` 與 `massif.XXXXX.ps` ，其中 `XXXXX` 是進程的 `PID`。`.txt` 檔案是所有調用點的記憶體使用摘要，而 `.ps` 檔案的內容如圖 7.7 。
```
process 是「行程」

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+必須使用 `-p IMGFILE` 參數來指定在 `PNG` 檔案 `IMGFILE`中產生圖形。收集資料的程式碼在程式中實際運行，不是像valgrind一樣使用模擬機。因此 `memusage` 的速度比 `massif` 快得多，並且可適用於 `massif` 無法使用的情況下。除了總記憶體消耗外，程式碼還記錄記憶體的分配大小，在程式結束時，它以直方圖顯示所有使用的記憶體分配大小。並把這些資料寫進標準錯誤(Stderr)。
+
+有些情況是不太能夠直接呼叫待觀察程式。舉個例子來說是 gcc 編譯器的編譯階段是由 gcc 驅動程式啟動的。在這種情況下使用 memusage 指令可以透過 -n NAME 提供被觀察程式的名稱，。就算被觀察的程式啟動了其他程式，也可以使用這個參數。如果沒有指定程式名稱，所有啟動的程式都會被記錄。
```
command 是「命令」，而非「指令」

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -0,0 +1,103 @@
+# 7.5 頁面錯誤最佳化
+
+在像Linux這樣具有需求分頁功能的作業系統上，`mmap` 的使用只修改頁面表。它確保對於檔案支持的頁面，可以找到底層的資料，對於匿名的記憶體空間，則提供初始化的頁面並將內容設定為0。`mmap` 使用時並不會實際分配記憶體。第一次讀取頁面時不管是讀寫資料或執行程式碼時，才會真的分配內存頁面的操作。當出現頁面錯誤時，作業系統核心會根據頁表樹決定需要在頁面上出現的資料，進行頁面錯誤解析。處理頁面錯誤的成本很高，但在進程使用每個記憶體頁面都時常發生。
```
留意 demand paging 的翻譯

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -0,0 +1,103 @@
+
+大多數架構還支援 1MB 或甚至更大的頁面大小。這樣的頁面在某些情況下也很有用，但這樣的記憶體分配法對實體RAM空間太過浪費。但大頁面也有其優點：如果使用巨大的資料集，以 x86-64 上的2MB頁面存儲它們比使用4k頁面節省了511個頁面錯誤（每個大頁面）。這可能會產生很大的差異。解決方法是有選擇地請求記憶體分配，僅某些地址範圍使用大型記憶體頁面，對於同一進程中的所有其他映射使用正常的頁面大小。
+
+使用大頁面的代價在於用於物理記憶體空間必須是連續的，因此存在內存碎裂的問題，可能在一段時間後就無法分配新的頁面空間。但處理記憶體碎片化是非常複雜的。例如對於2MB的大頁面，取得512個連續頁面窒礙難行，除非在系統啟動時。這就是為什麼目前大頁面的解決方案是使用特殊的文件系統 `hugetlbfs` 。可以通過寫入要保留的大頁面數量來做分配，檔案位置如下：
```
physical memory 的翻譯是「實體記憶體」，而非「物裡」

jserv reviewed (changes requested):
1. 注意中英文間的空白字元
2. 使用一致的詞彙

chiacyu committed:
Modify file content for enhancement

there're couple thing I did in this commit:

- fixed some typing error
- edited file content to be more readable
- enhancement based on reviewed messages

chiacyu commented:
I've made a couple of changes based on the reviewed message.

- attached the related link to tools
- fixed the typing error
- edited each sentence to reach higher readability
- half width space gap between each English characters and Chinese ones

Thanks!

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -0,0 +1,95 @@
+
+最簡單讀取頁面錯誤的方法是透過 time 工具。需要注意的是，要使用真正的工具，而不是 shell 的附給你的內建命令。如圖 7.4 中所示。
+
+<figure>
```
改用 Markdown 語法展現文字訊息

jserv commented on memory-performance-tools/simulating-cpu-caches.md:
```diff
@@ -0,0 +1,42 @@
+
+我們在 7.1 節所提到的 CPU 層面的分析工具像 oprofile 有助於我們了解快取的使用情況。其所有數據都是直接由硬體所收集，在不追求高精度的狀況下十分的好用。一旦牽涉需要高精度的資料時，oprofile 就派不上用場了；線程可能會被中斷非常多次。因此若是要查看程式在不同處理器上的記憶體使用行為，實際上必須湊齊這些不同的機台並直接在上面做實驗。但往往窒礙難行。像圖 3.8 的數據。若是要使用 oprofile 收集這些資料，必須要有 24 台不同的機器，但其中許多機器根本不存在。
+
+<figure>
```
改用 markdown 語法展現文字訊息

jserv commented on memory-performance-tools/simulating-cpu-caches.md:
```diff
@@ -0,0 +1,42 @@
+
+程式所產生的結果包含了快取使用摘要，以及每個函式對快取使用的詳細摘要。要獲得資料需要 `cg annotate` 能夠將記憶體位置與函式做配對，這表示最好要有除錯相關的資訊。若是沒有，雖然 `ELF` 檔案可以提供一些幫助，但是由於內部符號 (internal symbol) 沒有列在動態符號表中，因此結果不夠完整。圖 7.6 顯示了與圖 7.5 相同的程式運行的部分結果。
+
+<figure>
```
改用 markdown 語法展現文字訊息

chiacyu committed:
Replace certain png with markdown format

chiacyu commented:
Those reviewed messages have been implemented and pushed. 
Thanks!

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -0,0 +1,98 @@
+
+針對這個問題首先是避免觀測單一計數器的絕對值，可以將多個計數器的數值一起納入參考。處理器一次可以監控不只一個事件 (Event) ；然後可以計算收集到的絕對值的比值。可能可以獲得更好比較的結果。在計算時通常是以處理時間 (Process time) 作為除數，例如時脈週期數 (Clock cycles) 或指令數。以程式性能的初步測試而言，參考這兩個數字蠻有用的。
+
+圖 7.1 為對一個簡單隨機 “Follow” 測資進行量測的結果，縱軸為 CPI（Cycles Per Instruction）數值，橫軸則是不同的工作集 (Working set) 大小。在大多數 Intel 處理器上紀錄這個事件的變數名稱為 `CPU_CLK_UNHALTED` 和 `INST_RETIRED` 。從縮寫大概可以猜到，前者計算 CPU 的時脈週期，後者則計算指令數。我們看到了類似於我們使用的每個元素的週期量測的圖片。對於小工作集 (Working set) ，比值為 1.0 有時甚至會更低。這些計算結果是在 Intel Core 2 處理器上進行，該處理器具有多標量 (multi-scalar) 的特性，可以同時處理多個指令。對於不受記憶體頻寬限制的程式，比率可以顯著低於 1.0 ，但以這個案例來說，1.0 已經非常好了。
```
使用「測試資料」或「測試案例」，而非「測資」

jserv replied:
superscalar 的翻譯是「超純量」，multiscalar 的翻譯是「多純量」

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -0,0 +1,109 @@
+
+第二步驟是決定順序。各個目標檔案希望能使用最少的頁面數。作為額外的獎勵，沒有函式應該跨越頁面的邊界。在所有這些情況中的一個複雜之處是，為了最佳地安排目標檔案，必須知道鏈接器 (linker) 將之後的行為。在這裡鏈接器將按照它們出現在輸入檔（例如，壓縮）和命令行中的順序將目標檔案放入可執行文件或 DSO 中。對程式設計師來說相對可預期。
+
+對那些願意投入更多時間的人，可以使用自動追蹤呼叫的函式來重新排序，方法是使用 gcc 在 `-finstrument-functions` 選項時插入的 `__cyg_profile_func_enter` 和 `__cyg_profile_func_exit` 。詳細使用方式跟功能請參閱 gcc 使用手冊以獲取有關這些 `__cyg_*` 使用界面的更多資訊。通過追蹤程式的執行，程式設計師可以更準確地確定函式呼叫鏈。 [17] 的結果是通過重新排序函式來減少 5％ 的啟動成本。主要好處在於減少了頁面錯誤數量，但 TLB 快取也發揮了作用因為在虛擬化環境中，TLB 未命中的代價變得相對更加昂貴。
```
加上譯注:
> 這些函式之所以會用 `__cyg_` 開頭，是因收錄來自 Cygnus Solutions 公司這間早期的 GCC 貢獻者的程式碼，後來這間公司被 Red Hat 公司收購，其開發者持續投入 GCC 的維護，並維護前述函式的實作至今

jserv commented on memory-performance-tools/page-fault-optimization.md:
````diff
@@ -0,0 +1,109 @@
+一個防寫入的程式（所有程式都應該如此）可以使用以下函式在運行時確定掛載點存在：
+
+```c
+char *hugetlbfs_mntpoint(void) {
````
注意程式碼縮排

jserv commented on memory-performance-tools/simulating-cpu-caches.md:
```diff
@@ -0,0 +1,72 @@
+# 7.2 模擬處理器快取行為
+
+就算快取的運作原理相當容易理解，但實際程式究竟如何與快取互動並不容易觀察。其實程式設計師並不在乎資料所在的記憶體位置。記憶體位址產生可是由鏈接器 (Linker) 或在運行當下由動態鏈接器與作業系統決定。產生的組合語言能夠處理所有可能的記憶體位址。所以在程式碼中，甚至看不到任何關於記憶體位置的線索。也因為如此要理解程式如何使用記憶體真的非常的困難。
```
linker 的翻譯是「連結器」，而非「鏈結器」

jserv commented on memory-performance-tools.md:
```diff
@@ -0,0 +1,3 @@
+# 7. 記憶體性能量測工具
+
+有許多工具可供程式設計師使用來解程式的效能、快取和記憶體使用等等。現代處理器具備可用的效能監控硬體。但有些事件(Event)很難精確測量，因此需要模擬（Simulation）的空間。當牽涉到較高層次的功能時，有一些特殊工具可監控進程的執行。我們將介紹一組常用的工具，這些工具在大多數Linux系統上都可用。
```
用來解程式的效能 ==> 用以理解程式性能

jserv commented on memory-performance-tools/improving-branch-prediction.md:
```diff
@@ -0,0 +1,19 @@
+# 7.4 改善分支預測
+
+6.2.2 節中，我們提到了兩個改善 L1i 使用的方法，分別是通過分支預測和重新排序執行區塊：通過 `__builtin_expect` 進行靜態預測和狀態偵測的最佳化（PGO）。正確的分支預測對軟體的性能有巨大的影響，但在這裡我們會著重在記憶體使用的改善狀況。
```
通過 ==> 藉由

jserv commented on memory-performance-tools/improving-branch-prediction.md:
```diff
@@ -0,0 +1,19 @@
+
+6.2.2 節中，我們提到了兩個改善 L1i 使用的方法，分別是通過分支預測和重新排序執行區塊：通過 `__builtin_expect` 進行靜態預測和狀態偵測的最佳化（PGO）。正確的分支預測對軟體的性能有巨大的影響，但在這裡我們會著重在記憶體使用的改善狀況。
+
+使用 `__builtin_expect` （或 `likely` 和 `unlikely` 兩個巨集）的方式非常簡單。只要把相關的定義放在中央標頭檔裡，編譯器就可以發揮作用。但有個小問題，隨然對於軟體工程師來說，有可能在需要使用 `likely` 的狀況卻使用 `unlikely` ，反之亦然。即使可以使用類似 oprofile 的工具來測量分支預測錯誤和 L1i 未命中，但這些問題仍然很難抓漏。
```
這些問題仍然很難抓漏 => 這些問題不易排除

jserv commented on memory-performance-tools/improving-branch-prediction.md:
```diff
@@ -0,0 +1,19 @@
+
+PGO 對於 gcc 來說相當容易使用。總共只需要三個步驟，但必須滿足某些要求。首先，所有的來源檔案都必須使用額外的 `-fprofile-generate` 選項進行編譯。這個選項必須被傳遞到所有的編譯器運行的程式或鍊結該程式的命令中。雖然可以混合使用已啟用和未啟用此選項的檔案，但對於沒有啟用此選項的目標檔案，PGO 就不會被觸發。
+
+編譯器所產生的二進位檔案除了速度非常慢與容量巨大之外與正常的無異，因為他記錄（和儲存）有關分支是否被採用等等相關資訊。編譯器還為每個輸入的檔案新增一個名為 ".gcno" 的擴增檔案。這個檔案包含與程式碼中分支相關的資訊。這個檔案必須被保留以便後續使用。
```
他 => 它

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -0,0 +1,109 @@
+
+pagein 的輸出可用來確定程式碼中最理想相鄰順序。從 `/lib64/ld-2.5.so` 的程式碼可以發現，最初的指令呼叫了函式 `_dl_start` ，但這兩個函式位於不同的頁面上。重新排列程式碼會把程式碼移動到同一頁面以避免或延遲頁面錯誤。但要確認最佳程式碼順序是一個繁瑣的過程。由於設計上不會紀錄 pagein 不會紀錄第二次使用頁面的紀錄，因此要使用試錯法來查看更改的影響。並使用呼叫圖來分析，猜測可能的呼叫序列，有助於加快對函式和變數排序的過程。
+
+如果只是想有個粗略概念，可以通過查看組成可執行檔或動態函式庫的目標檔來查看呼叫序列。從單個或多個進入點（即函式名稱）開始，可以計算函式的互相依賴的關係。以目標檔案上來說這很容易。在每一輪中，確定包含所需函式和變量的目標文件。但種子 (seed) 集合必須事先指定。然後確定在這些目標文件中的所有未定義引用，加到所需符號集合中。重複執行直到集合穩定。
```
file 的翻譯是「檔案」，而非「文件」

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -0,0 +1,109 @@
+-  程式可以使用 System V 共享記憶體界面並設定 `SHM_HUGETLB` 標誌 (flag) 
+-  可以實際掛載 `hugetlbfs` 類型的檔案系統，然後程式可以在掛載點下新增一個檔案，並使用 `mmap` 將一個或多個頁面映射為記憶體空間。
+
+第一種情況下，`hugetlbfs` 不需要被掛載。
```
不需要被掛載 => 不用掛載
(流暢的漢語中，應免非出現必要的「被」)

jserv commented on memory-performance-tools/page-fault-optimization.md:
````diff
@@ -0,0 +1,109 @@
+void *a = shmat(id, NULL, 0);
+```
+
+這段程式碼的關鍵實做是使用 `SHM_HUGETLB` 旗標和選擇正確的 `LENGTH` ，這個值必須是系統的大頁面的倍數。不同的架構會有不同的值。使用 System V 共享記憶體介面有一個麻煩的問題，因為其透過 key 參數以區分（或共享）映射區域而且很容易跟 `ftok` 介面產生衝突因此最好使用其他機制。
````
implement 的翻譯是「實作」，見: https://hackmd.io/@sysprog/it-vocabulary

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -0,0 +1,109 @@
+  <figcaption>圖 7.9：Follow 在大頁面的輸出結果</figcaption>
+</figure>
+
+數字一開始比較小，但隨著工作集大小增加而開始遞增。使用大頁面的測試，在 512MB 的工作集上快了 38％ 。大頁面測試的曲線在大約 250 個週期處有形成一個高原。當工作集$2^{27}$個字元時數字再次顯著上升。高原形成的原因是 64 個 2MB 的 TLB 可以涵蓋$2^{27}$個字元。
```
改為 `當工作集 $2^{27}$ 個字元` (注意空白)

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -0,0 +1,109 @@
+
+正如這些數字所顯示，使用大工作集的成本中有很大一部分來自 TLB 未命中。也因此使用本節中描述的界面可能會帶來不錯的成果。圖表中的數字很可能是理想值的上限，但在現實世界中的某些應用程式也可以透過大頁面有顯著的加速效果。例如：資料庫，因為使用案例多是大量資料存取，是今天使用大頁面的應用程式之一。
+
+目前沒有辦法使用大頁面映射檔案支持 (filebacked) 的資料。但到目前為止提出的實做方式無非都是明確使用大頁面，並搭配使用 `hugetlbfs` 檔案系統。這種實做方式無法令人接受。大頁面的使用必須是透明，這樣作業系統核心才可以輕鬆地確定哪些映射適合並自動使用大頁面。問在於作業系統核心不會永遠知道使用模式。如果一個可以使用大頁面的記憶體空間要求 4k 頁面精細度（例如，因為使用 mprotect 更改了部分記憶體的保護範圍），在線性的實體記憶體空間則會浪費很多寶貴的資源。因此成功實現這種方法肯定需要更多時間。
```
留意 filebacked 的翻譯和前後文

chiacyu committed:
Modify file content
- Fixed some typing error
- Edited to be more readible
- Implemented reviewed message

jserv commented on memory-performance-tools/simulating-cpu-caches.md:
```diff
@@ -65,8 +65,8 @@ valgrind --tool=cachegrind \
 
-再次強調一下，`cachegrind` 是一個模擬器，不會使用來自處理器的測量數值。因此與處理器中實際的快取實做方式很可能會非常不同。`cachegrind` 使用 LRU 淘汰策略，但對於具有高關聯性的高速快取而言，這會造成額外的成本。除此之外，模擬器並沒有考慮 `context switch` 和系統呼叫 (System call) ，這兩者都可能破壞大部分 L2 快取並必須更新 L1i 和 L1d 的快取內容。這會導致模擬結果的總快取失誤數低於實際的狀況。儘管如此，`cachegrind` 還是一個很好的工具，可以了解程式的記憶體使用情況以及其相關的問題。
\ No newline at end of file
+再次強調一下，`cachegrind` 是一個模擬器，不會使用來自處理器的測量數據。因此與處理器中實際的快取實做方式很可能會非常不同。`cachegrind` 使用最近最少使用 (LRU) 策略，但對於具有高關聯性的高速快取而言，這會造成額外的成本。除此之外，模擬器並沒有考慮上下文交換 (Context switch) 和系統呼叫 (System call) ，這兩者都可能破壞大部分 L2 快取並必須更新 L1i 和 L1d 的快取內容。這會導致模擬結果的總快取失誤數低於實際的狀況。儘管如此，`cachegrind` 還是一個很好的工具，可以了解程式的記憶體使用情況以及其相關的問題。
```
"context switch" 和 "system call" 皆為小寫

jserv commented on memory-performance-tools/improving-branch-prediction.md:
```diff
@@ -0,0 +1,19 @@
+
+PGO 對於 gcc 來說相當容易使用。總共只需要三個步驟，但必須滿足某些要求。首先，所有的來源檔案都必須使用額外的 `-fprofile-generate` 選項進行編譯。這個選項必須被傳遞到所有的編譯器運行的程式或連結該程式的命令中。雖然可以混合使用已啟用和未啟用此選項的檔案，但對於沒有啟用此選項的目標檔案，PGO 就不會被觸發。
+
+編譯器所產生的二進位檔案除了速度非常慢與容量巨大之外與正常的無異，因為它記錄（和儲存）有關分支是否被採用等等相關資訊。編譯器還為每個輸入的檔案新增一個名為 ".gcno" 的擴增檔案。這個檔案包含與程式碼中分支相關的資訊。這個檔案必須被保留以便後續使用。
```
編譯器所產生的二進位檔案除了速度非常慢與容量巨大之外與正常的無異 ==> 開啟 PGO 後，除了編譯時間大幅增加外，編譯器產生的二進位檔案與尋常檔案無異
(通順且前後呼應)

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+有些程式有自己的記憶體配置程式或將系統的記憶體配置程式內嵌入至其他函式。第一種情況，分配器通常會直接忽略；在第二種情況下，記錄會因為系統的記憶體配置程式被多包一層導致一些資訊被隱匿，因為只有在最外層的函式呼叫位址會被記錄下來。因此最好將這些記憶體配置程式紀錄下來。參數 `--alloc-fn=xmalloc` 指定的 xmalloc 也是一種記憶體配置函式，這在 GNU 程式中很常見。但這只能將 xmalloc 的呼叫次數被記錄下來，但是 xmalloc 內部所做的呼叫不會被記錄。
+
+第二個工具叫 `memusage` ，來自 GNU C 的函式庫。可以說是 `massif` 的陽春版（但比 massif 早推出很多）。它只記錄堆 (Heap) 的總使用量（包括可能的 mmap 使用等等，如果使用 `-m`），也可以選擇紀錄堆疊 (stack) 。結果圖是隨時間變化的總記憶體使用情況，或者是記憶體分配函式使用的線性變化圖。這些圖表是由 `memusage` 腳本單獨新增，就像使用 valgrind 一樣，必須使用 memusage 腳本來啟動應用程式如下：
```
heap 的翻譯是「堆積」，stack 的翻譯是「堆疊」

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+必須使用 `-p IMGFILE` 參數來指定在 `PNG` 檔案 `IMGFILE` 中產生圖形。收集資料的程式碼會在程式中實際運行，而不像 valgrind 透過模擬機。因此 `memusage` 的速度比 `massif` 快得多，並且可適用於 `massif` 無法使用的情況下。除了總記憶體消耗外，程式碼還記錄記憶體的分配大小，在程式結束時，會以直方圖顯示所有使用的記憶體分配大小。並把這些資料寫進標準錯誤 (stderr) 。
+
+有些情況是不太能夠直接呼叫待觀察程式。舉個例子來說 gcc 編譯器的編譯階段是由 gcc 驅動程式啟動。在這種情況下使用 memusage 命令可以透過 -n NAME 提供被觀察程式的名稱，。就算被觀察的程式啟動了其他程式，也可以使用這個參數。如果沒有指定程式名稱，所有啟動的程式都會被記錄。
```
舉個例子來說 => 例如

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+圖中每一個方塊代表一個記憶體字元 (word) 。在比較小的記憶體區域中，有四個已分配的記憶體方塊。由於方塊內的標頭大小和對齊填充的開銷就佔 50％ 。因為標頭的放置方式會讓處理器的有效預取率降低 50％ 。如果按排列順序處理這些方塊（最適合預取），則處理器將讀取所有標頭和對齊填充字元到快取中，即使應用程式不應該讀取或寫入它們這些記憶體區塊。標頭字詞只有在運行中與之後釋放記憶體時才會使用。
+
+要改善這樣的情況其實可以改變實作方式，將管理資料的後設資料 (meta data) 放在其他地方。的確在某些狀況下可能會是個好主意。但是仍有許多事情需要考慮，其中安全性是個不可忽視的因素。不論未來有任何狀況，記憶體對齊所帶來的填充問題永遠不會消失（這個例子，如果忽略標頭則對齊填充佔了 16％ 的資料）。只有當程式設計師直接控制記憶體分配器時，才能避免這種情況。當對齊需求出現時，就算出現空洞，也在可控制範圍內。
+
```
去除非必要的空白行

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+Massif 還可以記錄程式的堆疊 (stack) 的使用情況，可以用來確定應用程式的總記憶體使用量，但並不適用於所有狀況。在某些情況下（線程堆疊或使用 signaltstack 時），`valgrind` 無法知道堆疊的使用限制，在這種例子就不會把這些堆疊數字加到總和中。還有一些其他情況，同樣也無法準確紀錄堆疊。如果程式屬於這些情況，可以在開始運行 massif 時將參數設定 --stacks=no 。要注意的是這些 `valgrind` 的使用參數，必須在加在觀察程式之前。
+
+有些程式有自己的記憶體配置程式或將系統的記憶體配置程式內嵌入至其他函式。第一種情況，分配器通常會直接忽略；在第二種情況下，記錄會因為系統的記憶體配置程式被多包一層導致一些資訊被隱匿，因為只有在最外層的函式呼叫位址會被記錄下來。因此最好將這些記憶體配置程式紀錄下來。參數 `--alloc-fn=xmalloc` 指定的 xmalloc 也是一種記憶體配置函式，這在 GNU 程式中很常見。但這只能將 xmalloc 的呼叫次數被記錄下來，但是 xmalloc 內部所做的呼叫不會被記錄。
```
allocate 的翻譯是「配置」，而非「分配」，這樣才可跟 dispatch (分派) 區分

jserv replied:
有些程式有自己的記憶體配置程式或將系統的記憶體配置程式內嵌入至其他函式 => 有些程式內建客製的記憶體配置器，或改寫系統提供的記憶體配置器

jserv commented on memory-performance-tools/page-fault-optimization.md:
````diff
@@ -0,0 +1,111 @@
+void *a = mmap(NULL, LENGTH,PROT_READ|PROT_WRITE,fd, 0);
+```
+
+透過在 `open` 呼叫中使用相同的檔名，多個行程可以共享相同的大頁面並共同協作。也可以將頁面設置為可執行，只是在 `mmap` 呼叫中必須設定 `PROT_EXEC`  標誌。與 System V 共享記憶體例子相同， `LENGTH` 的值必須是系統的大頁面大小的倍數。
````
並共同協作 => 並協作

jserv commented on memory-performance-tools.md:
```diff
@@ -0,0 +1,3 @@
+# 7. 記憶體性能量測工具
+
+有許多工具可供程式設計師使用來理解程式的效能、快取和記憶體使用等等。現代處理器具備可用的效能監控硬體。但有些事件(Event)很難精確測量，因此需要模擬（Simulation）的空間。當牽涉到較高層次的功能時，有一些特殊工具可監控進程的執行。我們將介紹一組常用的工具，這些工具在大多數 Linux 系統上都可用。
```
有許多工具可供程式設計師使用來理解程式的效能、快取和記憶體使用等等 => 許多工具可協助程式開發者理解程式效能、快取和記憶體使用狀況等資訊。

jserv replied:
event 和 simulation 都該為小寫

jserv commented on memory-performance-tools/improving-branch-prediction.md:
```diff
@@ -0,0 +1,19 @@
+
+當執行一組代表性測試後，就可以來重新編譯應用程式。編譯器必須能夠在包含來源檔案的同一目錄中找到 `.gcda` 檔。千萬不能移動這些檔案，否則編譯器會找不到，嵌入的檔案校驗也會無法匹配。重新編譯時，要將 `-fprofile-generate` 參數替換為 `-fprofile-use` 。最重要的是，原始碼更改有一些限制。更改空格和注釋沒問題，但如果加上更多分支或基本程式碼區塊會使收集的資料無效，導致編譯失敗。
+
+這些就是程式設計師所需要做；看起來很簡單對吧？最重要的是選擇最適合的工具來進行測量。如果測試工作不適合程式的實際使用方式，執行的結果可能會更糟。也因為如此，使用 PGO 來進行函式庫的最佳化通常非常困難。函式庫可以在許多情境中使用，除非使用案例相似，否則通常最好只使用 `__builtin_expect` 的靜態分支預測。
```
程式設計師 => 程式開發者

jserv replied:
這些就是程式設計師所需要做；看起來很簡單對吧？ => 這些就是程式開發者該有的準備，不難吧？

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+那麼如何知道什麼時候是使用 obstacks（或類似技術）的絕佳時機？如果沒有人可以問，真的非常的困難，但是圖表可以提供一些線索。如果從同一位置進行許多記憶體配置，這可能意味著大塊的記憶體配置可能有所幫助。在圖 7.7 中，我們可以在地址 `0x4c0e7d5` 的配置中看到這樣的可能選項。從 800 毫秒到 1800 毫秒，他是唯一一個記憶體增長的區域（除了綠色區域）。此外斜率不陡，這意味著我們有大量的小塊記憶體配置發生。這確實是使用 obstacks 或類似技術的絕佳時機。
+
+另一個問題是圖中顯示的是總分配次很高的情況。如果圖形不是按時間線性繪製，而是按呼叫次數線性繪製（ memusage 的原始設定），這就不容易看出來。在這種情況下，圖中的緩坡表示有很多小的記憶體分配。memusage 不會顯示記憶體分配發生的位置，但是可以通過與 massif 輸出的比較來確定，或者使用者可能會立即看出來。許多小的記憶體分配應該被合併以實踐線性記憶體分配。
```
注意空白字元

chiacyu committed:
Modify multiple pages content
- Implemented reviewed message
- Edited sentences to increase readability

jserv commented on memory-performance-tools/improving-branch-prediction.md:
```diff
@@ -0,0 +1,19 @@
+
+這些就是程式開發者該有的準備，不難吧？最重要的是選擇最適合的工具來進行測量。如果測試工作不適合程式的實際使用方式，執行的結果可能會更糟。也因為如此，使用 PGO 來進行函式庫的最佳化通常非常困難。函式庫可以在許多情境下使用，除非使用案例相似，否則最好只使用 `__builtin_expect` 的靜態分支預測。
+
+而對 `.gcno` 和 `.gcda` 兩個檔案而言，有幾點需要注意。這些二進位檔案並不能直接用來檢查。但可以使用 gcc 套件中的 gcov 工具來進行檢視。這個工具主要用於覆蓋分析（因此得名），但使用的檔案格式與 PGO 相同。 gcov 工具為每個包含執行程式碼的來源檔案（可能包括系統標頭）生成 `.gcov` 的擴增檔案。這些檔案是根據給定 gcov 的參數進行註釋與列表，其中包括分支計數器、機率等資料。
```
本段的原文是:
> A few words on the `.gcno` and `.gcda` files. These are binary files which are not immediately usable for inspection. It is possible, though, to use the gcov tool, which is also part of the gcc package, to examine them. This tool is mainly used for coverage analysis (hence the name) but the file format used is the same as for PGO. The gcov tool
generates output files with the extension .gcov for each source file with executed code (this might include system headers). The files are source listings which are annotated, according to the parameters given to gcov, with branch counter, probabilities, etc.

其中 annotate 若翻譯為「註釋」，可能會讓讀者誤會為 comment，但這二個動詞落差很大:
* [annotate](https://dictionary.cambridge.org/dictionary/english/annotate): to add a description or piece of information to data, for example a label saying whether a word is a noun, a verb, etc.
* [comment](https://dictionary.cambridge.org/dictionary/english/comment): something that you say or write that expresses your opinion

這裡的 annotate 著重於「夾帶用以日後分析用的資訊」，而非是「來自開發者的想法或意見」。

可將「這些檔案是根據給定 gcov 的參數進行註釋與列表，其中包括分支計數器、機率等資料」改為
> 這些檔案依據給定的 gcov 參數，對原始程式碼標注 (annotate) 分支計數器和程式碼執行的機率等資訊。

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+有些程式內建客製的記憶體配置器，或改寫系統提供的記憶體配置器。第一種情況，配置器通常會直接忽略；在第二種情況下，記錄會因為系統的記憶體配置程式被多包一層導致一些資訊被隱匿，因為只有在最外層的函式呼叫位址會被記錄下來。因此最好將這些記憶體配置程式紀錄下來。參數 `--alloc-fn=xmalloc` 指定的 xmalloc 也是一種記憶體配置函式，這在 GNU 程式中很常見。但這只能將 xmalloc 的呼叫次數被記錄下來，xmalloc 內部所做的呼叫不會被記錄。
+
+第二個工具叫 `memusage` ，來自 GNU C 的函式庫。可以說是 `massif` 的陽春版（但比 massif 早推出很多）。它只記錄堆積 (heap) 的總使用量（包括可能的 mmap 使用等等，如果使用 `-m`），也可以選擇紀錄堆疊 (stack) 。結果圖是隨時間變化的總記憶體使用情況，或者是記憶體配置函式使用的線性變化圖。這些圖表是由 `memusage` 腳本單獨新增，就像使用 valgrind 一樣，必須使用 memusage 腳本來啟動應用程式如下：
```
本段原文是
> The second tool is called memusage; it is part of the GNU
C library. It is a simplified version of massif (but existed
a long time before massif). It only records the total memory use for heap (including possible calls to mmap etc. if
the -m option is given) and, optionally, the stack. The
results can be shown as a graph of the total memory use
over time or, alternatively, linearly over the calls made to
allocation functions. The graphs are created separately
by the memusage script which, just as with valgrind, has
to be used to start the application:

可譯作:
> 第二個工具名為 `memusage`，是 GNU C 函式庫的一部分，可看作 `massif` 的簡化版本 (但早於 `massif` 許久)

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+- 由記憶體配置器管理的大型記憶體塊中的記憶體區塊實際上是從後面往前算
+- 原本的記憶體區塊用盡，並在定址空間索取在不同位置的新記憶體區塊
+- 配置請求的大小不同，從不同的記憶體池中提供服務
+- 多線程程式的各個線程交錯配置記憶體區塊
```
多執行緒程式

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+如果程式必須事先配置記憶體才能進行後續動作，使用鍊結串列顯然不適合。因為列表中連續項目在主記憶體中的儲存位置並不一定是按順序排列。如果要確保主記憶體上的順序，那就不能以小記憶體區塊進行配置。必須使用另一層記憶體處理程式，軟體工程師可以自己實作客製化的記憶體配置器。另一個選擇是使用 GNU C 函式庫中的 obstack 實作。這個配置器從系統的記憶體配置器請求大塊記憶體，再分成任意大小的小記憶體區塊。但這些配置區塊會呈連續排列，除非大記憶體區塊已用盡，主要取決於要求的大小但非常罕見。Obstacks 不能完全取代記憶體配置器，他在釋放記憶體的部份有其限制。詳細訊息，請參閱 GNU C 函式庫的使用說明書。
+
+那麼如何知道什麼時候是使用 obstacks（或類似技術）的絕佳時機？如果沒有人可以問，真的非常的困難。但是圖表提供一些線索。如果看到同一位置進行許多記憶體配置，這可能意味著一次大塊的記憶體配置可能有所幫助。在圖 7.7 中，我們可以在地址 `0x4c0e7d5` 的配置中看到這樣的情況。從800毫秒到1800毫秒，他是唯一一個記憶體增長的區域（除了綠色區域）。此外斜率不陡，這意味著我們有大量的小塊記憶體配置發生。這確實是使用 obstacks 或類似技術的絕佳時機。
```
本段原文是
> So, how can a situation where the use of obstacks (or similar techniques) is advisable be recognized from the graphs? Without consulting the source, possible candidates for the changes cannot be identified, but the graph can provide an entry point for the search. If many allocations are made from the same location, this could mean that allocation in bulk might help. In Figure 7.7, we can see such a possible candidate in the allocations at address 0x4c0e7d5. From about 800ms into the run until 1,800ms into the run this is the only area (except the
top, green one) which grows. Moreover, the slope is not steep, which means we have a large number of relatively small allocations. This is, indeed, a candidate for the use of obstacks or similar techniques.

可譯為:
> 那如何從圖表中識別出何時該用 obstacks 呢？若不查看原始程式碼，就無法確定哪些部分適合修改並運用，但圖表可提供搜索的進入點。

"obstack" 是 glibc 術語，可見: https://www.gnu.org/software/libc/manual/html_node/Obstacks.html
本文應使用小寫展現。

“technique" 不該翻譯為「技術」，這裡指手段或技巧。

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+  <figcaption></figcaption>
+</figure>
+
+圖中每一個方塊代表一個記憶體字元 (word) 。在比較小的記憶體區域中，有四個已配置的記憶體方塊。由於方塊內的標頭大小和對齊填充的開銷就佔50％ 。因為標頭的放置方式會讓處理器的有效預取率降低50％ 。如果按排列順序處理這些方塊（最適合預取），則處理器將讀取所有標頭和對齊填充字元到快取中，即便應用程式不應該讀取或寫入它們這些記憶體區塊。標頭字元只有在程式運行中與釋放記憶體時才會使用。
```
「記憶體字元」恐怕會造成誤解

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+圖中每一個方塊代表一個記憶體字元 (word) 。在比較小的記憶體區域中，有四個已配置的記憶體方塊。由於方塊內的標頭大小和對齊填充的開銷就佔50％ 。因為標頭的放置方式會讓處理器的有效預取率降低50％ 。如果按排列順序處理這些方塊（最適合預取），則處理器將讀取所有標頭和對齊填充字元到快取中，即便應用程式不應該讀取或寫入它們這些記憶體區塊。標頭字元只有在程式運行中與釋放記憶體時才會使用。
+
+要改善這樣的情況其實可以改變實作方式，將管理資料的後設資料 (meta data) 放在其他地方。在某些狀況下可能會是個好主意。但是仍有許多事情需要考慮，其中安全性是個不可忽視的因素。不論未來有任何狀況，記憶體對齊所帶來的填充問題永遠不會消失（這個例子，如果忽略標頭則對齊填充佔了16％的資料）。只有當程式開發者直接控制記憶體配置器時，才能避免這種情況。就算有對齊造成的空洞，也在可控制範圍內。
```
本段的原文是:
> One could at this point argue that the implementation should be changed to put the administrative data somewhere else. This is indeed done in some implementations, and it might prove to be a good idea. There are many aspects to be kept in mind, though, security not being the least of them. Regardless of whether we might see a change in the future, the padding issue will never go away (amounting to 16% of the data in the example, when ignoring the headers). Only if the programmer directly takes control of allocations can this be avoided. When alignment requirements come into play there can still be holes, but this is also something under control of the programmer.

原文沒提及 meta data

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -0,0 +1,111 @@
+# 7.5 頁面錯誤最佳化
+
+在像 Linux 這樣具有請求式的頁面調度系統 (demand paging) 的作業系統上，使用 `mmap` 只修改頁面表的內容。它確保對於檔案支持的頁面，可以找到底層的資料，對於匿名的記憶體空間，則提供初始化的頁面並將內容設定為0。`mmap` 使用時並不會實際配置記憶體。第一次讀取頁面時不管是讀寫資料或執行程式碼時，才會真的配置記憶體空間。當出現頁面錯誤時，作業系統核心會根據頁表樹決定需要在頁面上出現的資料，進行頁面錯誤處理。處理頁面錯誤的成本很高，但在行程使用每個記憶體頁面都時常發生。
```
demand paging 翻譯為「隨選分頁」。
[國家教育研究院](https://terms.naer.edu.tw/detail/52682e7a54c24097b06a33d3ced41b2e/)譯作「需求分頁」，這樣可能會讓讀者誤會為 "require pages"，但其實 demand paging 著重於「選擇什麼」就「給予什麼」，於是「隨選分頁」更合適。

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -0,0 +1,111 @@
+第二步驟是決定順序。各個目標檔案希望能使用最少的頁面數。作為額外獎勵，沒有函式應該跨越頁面的邊界。在所情況中的一個複雜之處是，為了最佳地安排目標檔案，必須知道連結器 (linker) 之後的行為。在這裡連結器將按照它們出現在輸入檔和命令行中的順序將目標檔案放入可執行檔或 DSO 中。對程式開發者來說相對可預期。
+
+對那些願意投入更多時間的人，可以使用自動追蹤呼叫的函式來重新排序，方法是使用 gcc 在 `-finstrument-functions` 選項時插入 `__cyg_profile_func_enter` 和 `__cyg_profile_func_exit` 。詳細使用方式跟功能請參閱 gcc 使用手冊以獲取有關這些 `__cyg_*` 使用界面的更多資訊。通過追蹤程式的執行，程式開發者可以更準確地確定函式呼叫鏈。 [17] 的結果是通過重新排序函式來減少5％的啟動成本。主要好處在於減少了頁面錯誤數量，但 TLB 快取也發揮了作用因為在虛擬化環境中，TLB 未命中的代價變得更加昂貴。
+> 這些函式之所以會用 __cyg_ 開頭，是因收錄來自 Cygnus Solutions 公司這間早期的 GCC 貢獻者的程式碼，後來這間公司 Red Hat 公司收購，其開發者持續投入 GCC 的維護，並維護前述函式的實作至今。
```
「譯註」的方式可見: https://raw.githubusercontent.com/sysprog21/cpumemory-zhtw/master/what-programmers-can-do/bypassing-the-cache.md

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -0,0 +1,111 @@
+
+Linux 系統核心提供了兩個額外的機制可以避免頁錯誤。第一個機制是 `mmap` 系統呼叫的標誌 (flag) ，可以告訴作業系統核心不要去修改頁面表的資料，但去對映射區域中的所有頁面先標記頁面錯誤。只需在 `mmap` 的第四個參數中加入 `MAP_POPULATE` 標誌即可。但這也會導致 `mmap` 的使用成本顯著增加，但如果這些映射空間的所有頁面都立即使用，好處會十分顯著。相較於一次處理多個頁面錯誤，每次處理一個頁面錯誤的成本可能更高，因為還需要處理同步化等等其他問題，與其如此不如讓程式只有一個昂貴的 `mmap` 呼叫。但使用這個標誌也有一些缺點，在映射頁面中如果有大部分的頁面在呼叫後沒有使用時，會造成時間和記憶體的浪費。預先處理的頁面如果沒有在短時間內被使用也會造成系統阻塞。由於記憶體在使用之前就已配置，這可能導致記憶體短缺。另一方面，在最糟的情況下，頁面可能會被重新使用 ...
+
+但 `MAP_POPULATE` 的方式有些過於粗糙。而且還有第二問題是實際上沒有必要將所有頁面都映射進來。如果系統太忙無法執行操作，已經預取的頁面也可以會被丟棄。一旦真的使用了該頁面，程式就會出現頁面錯誤。另一種方法是使用 `posix_madvise` 函式的 `POSIX_MADV_WILLNEED` 。這會提示作業系統程式在不久後會需要使用目標頁面。當然作業系統核心也可以忽略這個設定，但也可以預取先取好頁面。優點在於操作更細微，任何映射地址空間區域中的單個頁面或多個頁面範圍都可以被預先提取。對於許多在運行時不使用資料的記憶體映射檔案，可能比使用 `MAP_POPULATE` 效果還好。
```
將 `posix_madvise` 指向超連結 https://man7.org/linux/man-pages/man3/posix_madvise.3.html

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+現在我們已經知道如何捕獲有關記憶體配置的資料，要解讀這些資料就有必要在主記憶體和快取使用與互動來著手。有效的動態線性記憶體配置與高效率預取和減少快取未命中息息相關。
+
+需要讀取特定數量資料才能進行後續動作的程式可以通過新增一個列表來達成，其中每個列表項目都包含一個新的資料。這種配置方法的開銷可能是最小的（使用單向鍊結串列），但是在使用資料時的快取效果可能會大大降低性能。
```
linked list 翻譯為「鏈結串列」，注意「鏈」字

jserv commented on memory-performance-tools/page-fault-optimization.md:
````diff
@@ -0,0 +1,111 @@
+}
+```
+
+這兩種情況的更多使用資訊可以在作業系統核心樹的 hugetlbpage.txt 檔案中找到。該檔案還描述了 IA-64 所需的特殊處理方式。
````
改為:
> 關於這二種情況的更多資訊，可參見 Linux 核心原始程式碼內附文件 hugetlbpage.txt，該檔還描述 IA-64 特有處理機制。

追加 https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt 超連結

jserv commented on memory-performance-tools/simulating-cpu-caches.md:
```diff
@@ -0,0 +1,72 @@
+
+`cg annotate` 可以提供更詳細的資料。如果有特別指定某一個檔案，甚至可以逐行顯示出對應的快取命中數和未命中數。這些資訊可以讓程式開發者可以深入到確切行數。雖然程式介面有點粗糙：我在寫這篇文章時，cachegrind 資料檔案和原始檔必須在同一個目錄中。
+
+再次強調一下，`cachegrind` 是一個模擬器，不會使用來自處理器的測量數據。因此與處理器中實際的快取實作方式很可能會非常不同。`cachegrind` 使用最近最少使用 (LRU) 策略，但對於具高關聯性的高速快取而言，這會造成額外的成本。除此之外，模擬器並沒有考慮上下文交換 (context switch) 和系統呼叫 (system call) ，這兩者都可能破壞大部分 L2 快取且必須更新 L1i 和 L1d 的快取內容。這會導致模擬結果的總快取未命中數低於實際的狀況。儘管如此，`cachegrind` 還是一個很好的工具，可以了解程式的記憶體使用情況以及其相關的問題。
```
刪去「一下」，贅詞

jserv commented on memory-performance-tools/simulating-cpu-caches.md:
```diff
@@ -0,0 +1,72 @@
+
+當然 cachegrind 還可以支援更多事情。在行程結束之前，它會產生一個 `cachegrind.out.XXXXX` 檔，其中 `XXXXX` 是行程的 `PID` 。這個檔案包含了程式中每個函式和原始程式碼中快取使用的摘要和詳細內容。可以使用 `cg annotate` 來查看這些資料。
+
+程式所產生的結果包含了快取使用摘要，以及每個函式對快取使用的詳細摘要。要獲得資料需要 `cg annotate` 將記憶體位置與函式做配對，這表示最好要有除錯相關的資訊。若是沒有，雖然 `ELF` 檔案可以提供一些幫助，但是由於內部符號 (internal symbol) 沒有列在動態符號表中，因此結果不夠完整。圖 7.6 顯示了與圖 7.5 相同的程式運行的部分結果。
```
ELF 不用特別標示

jserv commented on memory-performance-tools/improving-branch-prediction.md:
```diff
@@ -0,0 +1,19 @@
+
+6.2.2 節中，我們提到了兩個改善 L1i 使用的方法，分別是藉由分支預測和重新排序執行區塊：藉由 `__builtin_expect` 進行靜態預測和狀態偵測的最佳化（PGO）。正確的分支預測對軟體的性能有巨大的影響，但在這裡我們會著重在記憶體使用的改善狀況。
+
+使用 `__builtin_expect` （或 `likely` 和 `unlikely` 兩個巨集）的方式非常簡單。只要把相關的定義放在中央標頭檔裡，編譯器就可以發揮作用。但有個小問題就是有可能在需要使用 `likely` 的狀況卻使用 `unlikely` ，反之亦然。即使可以使用類似 oprofile 的工具來測量分支預測錯誤和 L1i 未命中，但這些問題仍然不易排除。
```
「可以使用」都該簡化為「可用」，漢語可寫出緊湊的文字，翻譯時該善用

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -0,0 +1,97 @@
+
+針對這個問題首先是避免觀測單一計數器的絕對值，可以將多個計數器的數值一起納入參考。處理器一次可以監控不只一個事件；再來可以計算收集到的絕對值的比值。可能可以獲得更好比較的結果。在計算時通常是以處理時間 (process time) 作為除數，例如時脈週期數 (clock cycles) 或指令數。以程式性能的初步測試而言，這兩個數字具有參考的價值。
+
+圖 7.1 為對一個簡單隨機 “Follow” 測試資料進行量測的結果，縱軸為 CPI（Cycles Per Instruction）數值，橫軸則是不同的工作集 (Working set) 大小。在大多數 Intel 處理器上紀錄這個事件的變數名稱為 `CPU_CLK_UNHALTED` 和 `INST_RETIRED` 。從縮寫大概可以猜到，前者計算 cpu 的時脈週期，後者則計算指令數。我們看到了類似於我們使用的每個元素的週期量測的圖片。對於小工作集 (Working set) ，比值為 1.0 有時甚至會更低。這些計算結果是在 Intel Core 2 處理器上進行，此處理器具有多純量 (multi-scalar) 的特性，可以同時處理多個指令。對於不受記憶體頻寬限制的程式，比率可以顯著低於 1.0 ，但以這個案例來說，1.0 的結果已經非常好了。
```
Follow 是否該翻譯？

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -0,0 +1,97 @@
+* 大多數的預取指令沒有幫助
+* 預取指令應該要因硬體特性做調整
+
+這個部份就可以留給讀者去練習看看如何在硬體上調整出最佳解決方案。硬體規格扮演了至關重要的角色。對 Core 2 處理器而言，SSE 運算的延遲為1個時脈週期。若是較舊版本則為2個時脈週期，這表示硬體預取器和預取指令有更多的前置時間來讀取數據。
```
這裡加上「譯註」，x86 家族微處理器的指令延遲可見: https://www.agner.org/optimize/instruction_tables.pdf

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -0,0 +1,97 @@
+
+- 2. 並非所有事件都能被精準的記錄。舉例來說，在記錄特定事件時指令計數器的結果往往可能是錯誤的。處理器的多純量 (multi-scalar) 特性可能使計算結果很難100％正確。但在某些處理器上，某些事件可以精確紀錄。
+
+這些已標記的列表不僅有助於解讀預取資訊，每個事件都是以指令指標(pointer)來記錄，因此可以精準定位程式中的其他熱點。執行頻繁的位置通常會是許多 `INST_RETIRED` 事件的來源，因此值得進行額外的調整。如果觀測結果有許多快取未命中時，會需要透過預取指令來避免快取未命中。
```
instruction pointer 是完整的術語，不該拆為「指令指標(pointer)」，可改翻譯為「程式計數器」並在小括號中提及:
> 在 x86 家族的微處理器，稱為 instruction pointer，或簡稱 IP

https://en.wikipedia.org/wiki/Program_counter

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -0,0 +1,97 @@
+
+有一種事件可以在沒有硬體支援的情況下測量頁面錯誤 (page fault) 。作業系統負責解決頁面錯誤，在這些情況下，會把結果計算下來。並分成以下兩種類型的頁面錯誤：
+
+- **次要頁面錯誤** 對於未使用的匿名頁面（未被任何檔案紀錄使用）、寫時複製 (copy-on-write) 頁面以及內容已經在記憶體某處的其他頁面。
```
CoW 翻譯為「寫入時才複製」

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -0,0 +1,97 @@
+
+很明顯，主要頁面錯誤比次要頁面錯誤要付出更多的代價。但次要頁面錯誤所造成的損失也不容忽視。在任何情況下，都需要進入作業系統核心來進行操作，無論是尋找新的頁面，清除或寫入需要的資料，都必須修改頁表樹 (page table tree) 。最後還需要同步那些讀取或修改頁表樹的其他任務，並可能引發更多的延遲。
+
+最簡單讀取頁面錯誤的方法是透過 time 工具。需要注意的是，要使用真正的工具，而不是 shell 的內建命令。如下圖 7.4 中所示。
```
加入 time 的超連結: https://man7.org/linux/man-pages/man1/time.1.html
此處「真正的工具」翻譯會讓人誤會為，其他是「假的」工具，可改為「不是 shell 內建的 time 命令」

jserv commented on memory-performance-tools/memory-operation-profiling.md:
````diff
@@ -0,0 +1,97 @@
+最簡單讀取頁面錯誤的方法是透過 time 工具。需要注意的是，要使用真正的工具，而不是 shell 的內建命令。如下圖 7.4 中所示。
+
+```c
+$ \time ls /etc
````
修正命令，應為
```shell
$ time ls /etc
```

jserv commented on memory-performance-tools/memory-operation-profiling.md:
````diff
@@ -0,0 +1,97 @@
+int getrusage(__rusage_who_t who, struct rusage *usage)
+```
+
+參數 `who` 明確表示是哪個行程要求資訊。目前只有 `RUSAGE_SELF` 和 
````
thread 的翻譯是「執行緒」

chiacyu committed:
Add multiple hyper link to file

- Implemented suggestion from reviewed message
- Add hyper link to specific key word
- Removed redundant words from sentenses

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -1,47 +1,47 @@
+但是效能測量計數器應該要讓我們能更深入地了解處理器內部運作。因此我們需要思考處理器的實作細節。在這個章節，我們非常在乎快取處理的細節，因此必須查看與快取相關的事件。但這些事件以及相關名稱與計數方式都與處理器息息相關。這就是為什麼 oprofile 有其不便的地方，就算使用介面非常友善：程式開發者還是必須事先了解計數器相關詳細資料。在附錄 B 中會展示一些處理器的詳細資訊。
 
-對於 Core 2 處理器，所對應的事件分別為 `L1D_REPL` 、`DTLB_MISSES` 和 `L2_LINES_IN`。後者可以測量所有未命中以及由指令所造成的未命中，而非硬體預取 (prefetch) 所造成 。隨機 “Follow” 測試的結果如圖 7.2 所示。
+對於 Core 2 處理器，所對應的事件分別為 `L1D_REPL` 、`DTLB_MISSES` 和 `L2_LINES_IN`。後者可測量所有未命中以及由指令所造成的未命中，而非硬體預取 (prefetch) 所造成 。隨機 “跟隨” 測試的結果如圖 7.2 所示。
```
使用正確的中文引號 (即 `「` 和 `」`)

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -50,48 +50,51 @@ L2 快取未命中率一直保持為零，直到 L2 快取完全用盡為止；
+這些已標記的列表不僅有助於解讀預取資訊，每個事件都以程式計數器來記錄，因此可精準定位程式中的其他熱點。執行頻繁的位置通常會是許多 `INST_RETIRED` 事件的來源，因此值得進行額外調整。如果觀測結果有許多快取未命中時，會需要透過預取指令來避免快取未命中。
 
-有一種事件可以在沒有硬體支援的情況下測量頁面錯誤 (page fault) 。作業系統負責解決頁面錯誤，在這些情況下，會把結果計算下來。並分成以下兩種類型的頁面錯誤：
+有一種事件可在沒有硬體支援的情況下測量頁面錯誤 (page fault) 。作業系統負責解決頁面錯誤，在這些情況下會把結果計算下來。並分成以下兩種類型的頁面錯誤：
```
「把結果計算下來」不通順

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -50,48 +50,51 @@ L2 快取未命中率一直保持為零，直到 L2 快取完全用盡為止；
+對頁面錯誤影響最大階段在程式啟動階段。使用每個頁面都會產生頁面錯誤；特別是對 GUI 應用程式，使用的頁面越多，程式運作所需的時間與準備時間就越長。在 7.5 節中，我們會介紹一個專門用於測量初始化時間的工具。
 
-time 的底層透過 `rusage` 來運作。`wait4` 系統呼叫在父行程等待子行程終止時會寫入 `strcut rusage` ; 這樣的行為正好符合 time 的需求。但行程還是可以獲取自己的資源使用情況（正是 `rusage` 命名的由來）或已終止的子行程的資源使用資訊。
+time 透過 `rusage` 來運作。`wait4` 系統呼叫在父行程等待子行程終止時會寫入 `strcut rusage` ; 正好符合 time 的需求。但行程仍可獲取自身資源使用資訊（正是 `rusage` 命名由來）或已終止子行程的資源使用資訊。
```
避免父權主義的遺毒，本文將 parent process 翻譯為「親代行程」，而非「父行程」或「母行程」，不僅更中性，也符合英文原意。若寫作「父」，則隱含「母」的存在，但以行程的階層關係來說，沒有這樣成對關連性。若用「上代」會造成更多的混淆，在漢語中，「上一代」沒有明確的血緣關係 (例如「[炎黃子孫](https://dict.revised.moe.edu.tw/dictView.jsp?ID=155219)」與其說是血緣關係，不如說是傾向文化認同)，但「[親](https://dict.concised.moe.edu.tw/dictView.jsp?ID=25345)」的本意就是指名血緣和姻親關係。

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -50,48 +50,51 @@ L2 快取未命中率一直保持為零，直到 L2 快取完全用盡為止；
 
-對於想要透過減少頁面錯誤來增加程式效能的程式開發者可以常常查看這些資訊，收集分析並做出比較。從系統面來看，如果使用者擁有足夠系統權限，也可以透過 `/proc/<PID>/stat` 看到相關資訊。其中 `<PID>` 是想查看的行程 ID ，頁面錯誤的數值則位在第十到第十四欄。分別是行程及其子程的累進次要和主要頁面錯誤數。
\ No newline at end of file
+對於想透過減少頁面錯誤來增加程式效能的程式開發者可查看這些資訊，收集分析並做出比較。以系統面而言，如果使用者擁有足夠系統權限，也可透過 `/proc/<PID>/stat` 看到相關資訊。其中 `<PID>` 是欲查看的行程 ID ，頁面錯誤的數值則位在第十到第十四欄。分別是行程及其子程的累進次要和主要頁面錯誤數。
```
第十到第十四欄 => 第 10 到第 14 欄

jserv commented on memory-performance-tools/page-fault-optimization.md:
```diff
@@ -20,39 +20,38 @@
+如果只是想有個粗略概念，可通過查看組成可執行檔或動態函式庫的目標檔來查看呼叫序列。從單個或多個進入點（即函式名稱）開始，可計算函式的互相依賴的關係。以目標檔案 (object file) 上來說很容易。在每一輪開始之前先確定目標檔案是否包含所需函式和變數的。但種子 (seed) 集合必須明確指定。然後將這些目檔案中所有未定義的引用 (undefined reference) 都加到所需符號集合 (needed symbols)。重複執行直到集合穩定。
 
-第二步驟是決定順序。各個目標檔案希望能使用最少的頁面數。作為額外獎勵，沒有函式應該跨越頁面的邊界。在所情況中的一個複雜之處是，為了最佳地安排目標檔案，必須知道連結器 (linker) 之後的行為。在這裡連結器將按照它們出現在輸入檔和命令行中的順序將目標檔案放入可執行檔或 DSO 中。對程式開發者來說相對可預期。
+第二步驟是決定順序。各個目標檔案希望能使用最少的頁面數。作為額外獎勵，沒有函式應該跨越頁面的邊界。在所情況中的一個複雜之處是，為了最佳地安排目標檔案，必須知道連結器 (linker) 之後的行為。在這裡連結器將按照它們出現在輸入檔和命令行中的順序將目標檔案放入可執行檔或 [DSO](https://en.wikipedia.org/w/index.php?title=Dynamic_Shared_Object&redirect=no) 中。對程式開發者來說相對可預期。
```
用更短的網址: https://en.wikipedia.org/wiki/Dynamic_linker

chiacyu committed:
Modify multiple pages
- Implemented reviewed messages
- Removed redundant punctuations
- Shorten certain sentance to be more readable

jserv commented on memory-performance-tools/improving-branch-prediction.md:
```diff
@@ -2,18 +2,18 @@
 雖然如此仍有一個簡單的方法可使用。A.2 節的程式碼另外定義了 `likely` 和 `unlikely` 兩個巨集，可在程式執行階段主動測量靜態預測是否正確。且使用者或測試人員可就檢查結果進行調整。但測量不會考量程式的性能，只會測試程式設計師所做的靜態假設。詳細內容以及程式碼可在先前的章節中找到。
 
-[PGO](https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-8/profile-guided-optimization-pgo.html) 對 [gcc](https://www.man7.org/linux/man-pages/man1/gcc.1.html) 來說相當容易使用。只需要三個步驟，但須滿足某些先決條件。首先所有來源檔案都必須使用額外 [`-fprofile-generate`](https://gcc.gnu.org/onlinedocs/gcc-11.2.0/gcc/Instrumentation-Options.html) 選項進行編譯。這個選項必須被傳遞到所有編譯器運行的程式或連結該程式的命令中。雖然可混合使用已啟用和未啟用此選項的檔案， ...
+[PGO](https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-8/profile-guided-optimization-pgo.html) 對 [gcc](https://www.man7.org/linux/man-pages/man1/gcc.1.html) 來說相當容易使用。只需要三個步驟，但須滿足某些先決條件。首先所有來源檔案都須使用額外 [`-fprofile-generate`](https://gcc.gnu.org/onlinedocs/gcc-11.2.0/gcc/Instrumentation-Options.html) 選項進行編譯。這個選項必須被傳遞到所有編譯器運行的程式或連結該程式的命令中。雖然可混合使用已啟用和未啟用此選項的檔案，但 ...
```
改用 https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html 這個超連結，避免限定在特定的 GCC 版本

chiacyu committed:
Replace hyper link version

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -41,9 +41,9 @@ massif 和 memusage 都有提供額外的選項。程式開發者應先查閱使
+如果程式必須事先配置記憶體才能進行後續動作，使用鏈結串列顯然不適合。因為列表中連續項目在主記憶體中的儲存位置並不一定按順序排列。如果要確保主記憶體上的順序，就不能以小記憶體區塊進行配置。必須使用另一層記憶體處理程式。程式開發者可自己實作客製化記憶體配置器。另一個選擇是使用 GNU C 函式庫中的 [obstacks](https://www.gnu.org/software/libc/manual/html_node/Obstacks.html) 實作。這個配置器先從系統的記憶體配置器請求大塊記憶體，再分成任意大小的小記憶體區塊。但這些配置區塊會呈連續排列，除非大記憶體區塊已用盡，主要取決於要求的大小但非常罕見。obstacks 不能完全取代記憶體配置器，其釋放記憶體的部份有限制。詳細資訊可參閱 GNU C 函式庫的使用說明書。
 
-那如何從圖表中識別出何時該用 obstacks 呢？若不查看原始程式碼，就無法確定哪些部分適合修改並運用，但圖表可提供搜索的進入點。如果看到同一位置進行許多記憶體配置，可能表示一次大塊的記憶體配置可能有所幫助。在圖 7.7 中，我們可在地址 `0x4c0e7d5` 的配置中看到這樣的情況。從800毫秒到1800毫秒，他是唯一一個記憶體增長的區域（除了綠色區域）。此外斜率不陡，這表示有大量的小塊記憶體配置發生。確實是使用 obstacks 或類似技術的絕佳時機。
+那如何從圖表中識別出何時該用 obstacks 呢？若不查看原始程式碼，就無法確定哪些部分適合修改並運用，但圖表可提供搜索的進入點。如果看到同一位置進行許多記憶體配置，可能表示一次大塊的記憶體配置可能有所幫助。在圖 7.7 中，我們可在地址 `0x4c0e7d5` 的配置中看到這樣的情況。從800毫秒到1800毫秒，他是唯一一個記憶體增長的區域（除了綠色區域）。此外斜率不陡，表示有大量的小塊記憶體配置發生。確實是使用 obstacks 或類似技術的絕佳時機。
```
「從800毫秒到1800毫秒」 => 「從 800 毫秒到 1800 毫秒」
(注意空白)

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -19,7 +19,7 @@ Oprofile 提供簡單的界面，即便選擇 GUI 仍可運行於裝置底層。
 但是效能測量計數器應該要讓我們能更深入地了解處理器內部運作。因此我們需要思考處理器的實作細節。在這個章節我們非常在乎快取處理的細節，因此必須查看與快取相關的事件。但這些事件以及相關名稱與計數方式都與處理器息息相關。這就是為什麼 oprofile 有其不便，就算使用介面非常友善：程式開發者還是必須事先了解計數器相關詳細資料。在附錄 B 中會展示一些處理器的詳細資訊。
 
-對於 Core 2 處理器，所對應的事件分別為 `L1D_REPL` 、`DTLB_MISSES` 和 `L2_LINES_IN`。後者可測量所有未命中以及由指令所造成的未命中，而非硬體預取 (prefetch) 所造成 。隨機「跟隨」測試的結果如圖 7.2 所示。
+對於 [Core 2](https://oprofile.sourceforge.io/docs/intel-core2-events.php) 處理器，所對應的事件分別為 `L1D_REPL` 、`DTLB_MISSES` 和 `L2_LINES_IN`。後者可測量所有未命中以及由指令所造成的未命中，而非硬體預取 (prefetch) 所造成 。隨機「跟隨」測試的結果如圖 7.2 所示。
```
將 Core 2 改為 Intel Core 2，因為這已相當老舊

jserv commented on memory-performance-tools/memory-operation-profiling.md:
```diff
@@ -80,7 +80,7 @@ $ time ls /etc
 圖 7.4：time 工具的相關內容
 
-注意圖中最後一行。time 顯示1個主要頁面錯誤的和335個次要的頁面錯誤。確切數字可能不同；特別是如果立刻重複運行可能會顯示現在完全沒有主要的頁面錯誤。如果程式執行相同操作，在系統環境沒有任何變化情況下，總頁面錯誤的數字保持穩定。
+注意圖中最後一行。time 顯示1個主要頁面錯誤的和335個次要的頁面錯誤。確切數字可能不同；特別是如果立刻重複運行可能會顯示現在完全沒有主要的頁面錯誤。如果程式執行相同操作，在系統環境沒有任何變化情況下，總頁面錯誤的數字會保持穩定。
```
「time 顯示1個主要頁面錯誤的和335個次要的頁面錯誤」 => 「time 指出 1 個主要頁面錯誤的和 335 個次要的頁面錯誤」
(注意空白和動詞)

jserv replied:
「總頁面錯誤的數字會保持穩定」 => 「總頁面錯誤的數值保持穩定」

chiacyu committed:
Update file content

- Added more hyper link to certain keyword
- Implemented the suggestions of reviewed message

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+需要讀取特定數量資料才能進行後續動作的程式可通過新增一個列表來改善效能，其中每個列表項目都包含一個新的資料。這種配置方法的開銷可能是最小的（使用單向鏈結串列），但使用資料時的快取效果可能會大大降低性能。
+
+其中一個問題，按時序配置的記憶體區塊在主記憶體 (RAM) 上並非按順序排列。有許多可能的原因：
```
移除 `RAM`，此術語在第二章已有原文對照

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+圖中每一個方塊代表一個記憶體[字 (word)](https://en.wikipedia.org/wiki/Word_(computer_architecture)) 。在比較小的記憶體區域中，有四個已配置的記憶體方塊。由於方塊內的標頭大小和對齊填充開銷就佔50％ 。因為標頭的放置方式會讓處理器的有效預取率降低50％ 。如果按排列順序處理這些方塊（最適合預取），處理器會讀取所有標頭和對齊填充字到快取中，即便應用程式不應該讀取或寫入它們這些記憶體區塊。標頭字只有在程式運行中與釋放記憶體時才會使用。
+
+要改善這樣的情況其實可以改變實作方式，將管理資料的後設資料 (administrative data) 放在其他地方。在某些狀況可能是個好主意。但是仍有許多事情需要考慮，其中安全性是不可忽視的因素。不論未來有任何狀況，記憶體對齊所造成的填充問題永遠不會消失（如果忽略標頭則對齊填充佔了16％的資料）。只有當程式開發者直接控制記憶體配置器時，才能避免這種情況。就算有對齊造成的空洞，也在可控範圍內。
```
administrative data 的翻譯不是「後設資料」，後者指 metadata。原文出現三次該詞彙。

原文:
> One could at this point argue that the implementation
should be changed to put the administrative data somewhere else.

可譯作「將維護記憶體操作所需的資料置於他處」

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+  <figcaption></figcaption>
+</figure>
+
+圖中每一個方塊代表一個記憶體[字 (word)](https://en.wikipedia.org/wiki/Word_(computer_architecture)) 。在比較小的記憶體區域中，有四個已配置的記憶體方塊。由於方塊內的標頭大小和對齊填充開銷就佔50％ 。因為標頭的放置方式會讓處理器的有效預取率降低50％ 。如果按排列順序處理這些方塊（最適合預取），處理器會讀取所有標頭和對齊填充字到快取中，即便應用程式不應該讀取或寫入它們這些記憶體區塊。標頭字只有在程式運行中與釋放記憶體時才會使用。
```
注意 `50%` 前方應有一個半形空白字元

jserv commented on memory-performance-tools/measuring-memory-usage.md:
```diff
@@ -0,0 +1,61 @@
+
+在行程終止前，massif 會新增兩個檔案分別為：`massif.XXXXX.txt` 與 `massif.XXXXX.ps` ，其中 `XXXXX` 是行程的 `PID`。`.txt` 檔案是行程所有呼叫點的記憶體使用摘要，而 `.ps` 檔案內容如圖 7.7 。
+
+massif 還可記錄程式堆疊 (stack) 的使用情況，可用來確定應用程式的總記憶體使用量，但並不適用於所有狀況。在某些情況下（執行緒的堆疊或使用 signaltstack 時），`valgrind` 無法知道堆疊的使用限制，在這個例子就不會把這些堆疊數字加到總和中。還有一些情況，同樣也無法準確紀錄堆疊。如果程式屬於這些情況，可在開始運行 massif 時將參數設定 --stacks=no 。要注意是這些 `valgrind` 的使用參數，必須加在觀察程式之前。
```
在本書 4.2 節已提及 stack 和 heap 原文，在此應該直接書寫譯文，不用提及對應的原文術語。

jserv replied:
標示 `--stacks=no`

jserv commented on memory-performance-tools/simulating-cpu-caches.md:
```diff
@@ -0,0 +1,72 @@
+
+`Ir` 、`Dr` 和 `Dw` 這幾個欄位顯示的是總快取使用量，而非快取未命中數，這些未命中次數會在後面兩欄顯示。這些數據可用於辨識出發生最多快取未命中的程式碼。首先可先關注 `L2` 的快取未命中，再延伸處理 `L1i/L1d` 快取未命中。
+
+`cg annotate` 可提供更詳細的資料。如果有特別指定某一個檔案，甚至可逐行顯示出對應的快取命中數和未命中數。這些資訊可讓程式開發者深入到確切行數，雖然使用介面粗糙。我在寫這篇文章時，cachegrind 資料檔案和原始檔必須在同一個目錄中。
```
我在寫這篇文章時 => 本文書寫之際

jserv commented on memory-performance-tools/page-fault-optimization.md:
````diff
@@ -0,0 +1,110 @@
+void *a = shmat(id, NULL, 0);
+```
+
+這段程式碼的關鍵實作是使用 `SHM_HUGETLB` 旗標和選擇正確的 `LENGTH` ，這個值必須是系統的大頁面的倍數。不同的架構會需要填入不同的值。使用 System V 共享記憶體介面有一個麻煩的問題，因為需要透過 key 參數以區分（或共享）映射區域而且很容易跟 `ftok` 介面產生衝突因此最好使用其他機制。
````
加入對應的超連結: https://man7.org/linux/man-pages/man3/ftok.3.html

jserv commented on memory-performance-tools/page-fault-optimization.md:
````diff
@@ -0,0 +1,110 @@
+/proc/sys/vm/nr_hugepages
+```
+
+如果無法找到足夠連續的記憶體空間，操作可能會失敗。在使用虛擬化技術時，情況會變得特別有趣。由於 VMM 模型的虛擬系統並不能直接管理實體的記憶體空間，因此無法自行配置 `hugetlbfs` 。必須依賴 VMM 代為處理，但這個功能並不是所有虛擬機都有。對於 KVM 模型，運行 KVM 模組的 Linux 作業系統核心可執行 `hugetlbfs` 配置，並可能將其中的一部分頁面傳遞給其中一個客戶空間。
````
增加譯註:
> VMM 是 virtual machine manager 或 virtual machine monitor 的簡稱，詳見 Wikipedia [hypervisor](https://en.wikipedia.org/wiki/Hypervisor) 詞目

chiacyu committed:
Update file content

jserv commented:
Thank @chiacyu for contributing!
